// @aws-cdk/pipelines/lib/codepipeline/codebuild-step.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeBuildStep = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const blueprint_1 = require("../blueprint");
/**
 * Run a script as a CodeBuild Project.
 *
 * @stability stable
 */
class CodeBuildStep extends blueprint_1.ShellStep {
    /**
     * @stability stable
     */
    constructor(id, props) {
        super(id, props);
        this.projectName = props.projectName;
        this.buildEnvironment = props.buildEnvironment;
        this.partialBuildSpec = props.partialBuildSpec;
        this.vpc = props.vpc;
        this.subnetSelection = props.subnetSelection;
        this.role = props.role;
        this.rolePolicyStatements = props.rolePolicyStatements;
        this.securityGroups = props.securityGroups;
    }
    /**
     * CodeBuild Project generated for the pipeline.
     *
     * Will only be available after the pipeline has been built.
     *
     * @stability stable
     */
    get project() {
        if (!this._project) {
            throw new Error('Call pipeline.buildPipeline() before reading this property');
        }
        return this._project;
    }
    /**
     * The CodeBuild Project's principal.
     *
     * @stability stable
     */
    get grantPrincipal() {
        return this.project.grantPrincipal;
    }
    /**
     * Set the internal project value
     *
     * @internal
     */
    _setProject(project) {
        this._project = project;
    }
}
exports.CodeBuildStep = CodeBuildStep;
_a = JSII_RTTI_SYMBOL_1;
CodeBuildStep[_a] = { fqn: "@aws-cdk/pipelines.CodeBuildStep", version: "1.121.0" };
